### ISSUE TITLE
Decentralize queries

### ISSUE THREAD

**Issue by cjgillot** [rust-lang/compiler-team#277]
# TL;DR

Allow each compiler crate to define its own queries, instead of declaring everything in librustc_middle.
This would remove the need for datatype definitions in librustc_middle.

# Links and Details

Exploratory work: moving the query system out of librustc_middle -> https://github.com/rust-lang/rust/pull/70951

# Mentors or Reviewers

I have not looked for people yet.
@eddyb might be interested because of https://github.com/rust-lang/rust/pull/70951#issuecomment-612156842

----

**Note:** Discussion of the proposal should happen on the associated [zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Decentralize.20queries.20compiler-team.23277/near/194545129) (read access available at associated [zulip archive](https://zulip-archive.rust-lang.org/233931tcompilermajorchanges/16971Decentralizequeriescompilerteam277.html)). This issue is meant to summarize the high level points and to track status of the proposal.

**Comment by nikomatsakis** [comment]
@cjgillot heh I can't find the zulip stream ... so I like the *idea* here but I feel like we're missing some details of how you plan to go about it. It's a big PR -- can you maybe update the description with some details of how you are decentralizing exactly?

**Comment by nikomatsakis** [comment]
Closing per @cjgillot's [comment on Zulip](https://zulip-archive.rust-lang.org/233931tcompilermajorchanges/74859Movequeryimplementationoutsiderustcmcompilerteam388.html#217279143).
